SUIT CI tests DON'T MERGE

kaspar030 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds CI testing to examples/suit_update.
Based on https://github.com/future-proof-iot/RIOT/pull/8.

The PR includes some dependencies already PR'ed to master, so it should not be merged.

The test script requires the flashed APP_VER to be 0 before starting. This might break the examples, and needs to be checked.

Other than that, "make test" using sudo start_network.sh doesn't work at all for me. @fjmolinas I guess it did for you?

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure


<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 committed:
riotboot: set FLASHFILE to RIOTBOOT_EXTENDED_BIN

... if the riotboot feature is used.

Previously, even an application that had "FEATURES_REQUIRED += riotboot"
set would still flash the non-riotboot binary on "make flash".
This is usualy not what the user wants.

This commit set's the FLASHFILE variable to the combined "riotboot
bootloader + slot0 + empty slot1" binary. This has the effect that make
all, flash and flash-only will compile and/or flash a working riotboot
setup.

tests/riotboot and tests/riotboot_flashwrite now default to flashing the
riotboot-extended binary. tests/riotboot was previously configured to
use the riotboot-combined binary. This has been changed in order to not
behave differently than how usual riotboot applications do.

kaspar030 committed:
riotboot: update documentation to match automatic use

FLASHFILE is now set to RIOTBOOT_EXTENDED_BIN, changing the meaning of
make all`, `flash`, `flash-only`.
This commits updates the documentation accordingly.

kaspar030 committed:
riotboot: fix APP_VER default documentation

kaspar030 committed:
tests/riotboot_flashwrite: disable for saml1[01]-xpro

The riotboot build fails because it only offers half the flash size (per
slot).

kaspar030 committed:
murdock: allow multiple files to be sent along with a test job

Previously, this was hard-coded to allow one file, hard-coded to be
called "flash file".
This commit allows multiple files to be specified via adding them to the
TEST_EXTRA_FILES variable. All files will be stored in the worker's
application bin directory.

Also, the existence check has been removed, as dwqc bails out on missing
file anyways.

kaspar030 committed:
tests/riotboot: add re-flashing and image selection test

kaspar030 committed:
make/riotboot: allow slot image creation without compiling

kaspar030 committed:
fixup! riotboot: fix APP_VER default documentation

kaspar030 committed:
fixup! tests/riotboot: add basic automatic test in python

kaspar030 committed:
riotboot: reference riotboot test documentation

Also, drop outdated manual testing procedure.

kaspar030 committed:
riotboot: always link both slot .elf when linking

This enables "make test" and "make test-murdock" to rely on both images being
present and thus execute without re-compiling.

kaspar030 committed:
fixup! riotboot: always link both slot .elf when linking

kaspar030 committed:
fixup! riotboot: set FLASHFILE to RIOTBOOT_EXTENDED_BIN

kaspar030 committed:
fixup! murdock: allow multiple files to be sent along with a test job

kaspar030 committed:
fixup! fixup! murdock: allow multiple files to be sent along with a test job

kaspar030 committed:
fixup! fixup! fixup! murdock: allow multiple files to be sent along with a test job

kaspar030 committed:
examples/suit_update: add automatic test script, cleanups

Test script was slightly adapted from @fjmolina's work.

Co-authored-by: Francisco Molina <femolina@uc.cl>

kaspar030 commented:
> The test script requires the flashed APP_VER to be 0 before starting. This might break the examples, and needs to be checked.

This I don't know yet how to work around. CI and test script need APP_VER=0. For users, APP_VER=$epoch is nicest. Would be a shame to have an extra test application just to acomodate this difference.

kaspar030 commented:
The branch contains some newer master commits and some master PR dependencies. Only the two topmost commits should be reviewed here.

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,124 @@
+        tmpdir, aiocoap_process = start_aiocoap_fileserver()
+        # TODO: wait for coap port to be available
+
+        publish(tmpdir.name, COAP_HOST, 1)
```
```
publish(tmpdir.name, COAP_HOST, 1, 1)
publish(tmpdir.name, COAP_HOST, 2, 2)
```

fjmolinas commented on examples/suit_update/Makefile:
```diff
@@ -39,61 +39,77 @@ USEMODULE += shell_commands
+# The test script assumes initially version "0" is flashed,
+# as done by the CI.
+# Thus default to that (instead of epoch set by makefiles/boot/riotboot.inc.mk).
+APP_VER ?= 0
```
Could be removed, see above comment.

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,124 @@
+def start_aiocoap_fileserver():
+    tmpdirname = tempfile.TemporaryDirectory()
+    aiocoap_process = subprocess.Popen(
+        "exec aiocoap-fileserver %s" % tmpdirname.name, shell=True
```
It needs to be documented that this was added to your path, I thinks that what you did right?

kaspar030 replied:
yup

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def cleanup(tmpdir, aiocoap_process):
+    aiocoap_process.kill()
```
This doesn't kill everything properly, specificity fails to clean up the created `tap` interface. SIGKILL is killing all child processes but not removing the interface, this is because in the script:

`trap "cleanup" INT QUIT TERM EXIT`.

Can you tell if there is a reason for this? Or should we push the change in master? I'll take a look at the history.

kaspar030 replied:
Ah, the shell script cannot trap "KILL". Need to change to send "TERM" first, then "KILL" a little later...

kaspar030 replied:
sorry, this was outdated, with setup_network.sh, this isn't an issue anymore

fjmolinas reviewed (changes requested):
What does it have to do with RIOT-OS#31 "migrate native to current Makefile structure" ?

> Other than that, "make test" using sudo start_network.sh doesn't work at all for me. @fjmolinas I guess it did for you?

Are you doing `sudo make test`? Or running it separately?

Anyway this has changes my setup a lot, publishing is done in the same go so it is running with sudo and publishing files with sudo. This means everything needs to be killed with sudo too, and `process.kill()` doesn't cleanup everything `start_network.sh` sets up, so I usually have to call test twice, once for cleanup and another to setup and tun.

> This I don't know yet how to work around. CI and test script need APP_VER=0. For users, APP_VER=$epoch is nicest. Would be a shame to have an extra test application just to accommodate this difference.

Do you really need `APP_VER=0`? I bodged a workaround, a gist [HERE](https://gist.github.com/fjmolinas/ba517d4407905e3026fe5bbb0b99719d).

Anyway it works for me with or without the patch but only for the first round of update, it is always using "latest" so the second try fails. When publishing it did specify a `latest_name` to easily pick the fw to notify, if you change the publish to:

```
        publish(tmpdir.name, COAP_HOST, 1, 1)
        publish(tmpdir.name, COAP_HOST, 2, 2)
```

and `SUIT_NOTIFY_VERSION` to `SUIT_MANIFEST_SIGNED_LATEST` for notifying it works. `SUIT_NOTIFY_VERSION` is not in the makefile so does nothing, you could also set `SUIT_NOTIFY_VERSION ?= SUIT_MANIFEST_SIGNED_LATEST' and change the notify recipe to:

```
suit/notify: | $(filter suit/publish, $(MAKECMDGOALS))
	@test -n "$(SUIT_CLIENT)" || { echo "error: SUIT_CLIENT unset!"; false; }
	aiocoap-client -m POST "coap://$(SUIT_CLIENT)/suit/trigger" \
		--payload "$(SUIT_COAP_ROOT)/$$(basename $(SUIT_NOTIFY_VERSION))" && \
		echo "Triggered $(SUIT_CLIENT) to update."
```

Doing this changes the test works for me. Here is my diff:

```
index 2b5986853..3be2de5f4 100755
--- a/examples/suit_update/tests/01-run.py
+++ b/examples/suit_update/tests/01-run.py
@@ -45,7 +45,7 @@ def notify(coap_server, client_url, version=None):
         "SUIT_CLIENT={}".format(client_url),
     ]
     if version is not None:
-        cmd.append("SUIT_NOTIFY_VERSION={}".format(version))
+        cmd.append("SUIT_MANIFEST_SIGNED_LATEST={}".format(version))
     assert not subprocess.call(cmd)
 
 
@@ -111,8 +111,8 @@ if __name__ == "__main__":
         tmpdir, aiocoap_process = start_aiocoap_fileserver()
         # TODO: wait for coap port to be available
 
-        publish(tmpdir.name, COAP_HOST, 1)
-        publish(tmpdir.name, COAP_HOST, 2)
+        publish(tmpdir.name, COAP_HOST, 1, 1)
+        publish(tmpdir.name, COAP_HOST, 2, 2)
 
```

And the command sequence I ran is:

```
make -C examples/suit_update/ BOARD=samr21-xpro -j3 flash

sudo make -C examples/suit_update/ BOARD=samr21-xpro test
```

PS: small comment. I rather have Francisco Molina as naming in the commit co-authorship (and for all future cases).

kaspar030 committed:
examples/suit_update: update README.md

kaspar030 commented:
> What does it have to do with [RIOT-OS#31](https://github.com/RIOT-OS/RIOT/pull/31) "migrate native to current Makefile structure" ?

Messup, I meant your PR: future-proof-iot/RIOT#8

kaspar030 commented:
>  small comment. I rather have Francisco Molina as naming in the commit co-authorship (and for all future cases).

fixed

kaspar030 committed:
dist/tools/ethos: add setup_network.sh script

kaspar030 committed:
suit: allow specifying which image to use for suit/notify

kaspar030 committed:
examples/suit_update: change to make use of setup_network.sh

kaspar030 committed:
examples/suit_update: drop APP_VER?=0 requirement of test script

fjmolinas commented:
@kaspar030 I see you pushed a lot of changes, should I re-test? or are more changes coming?

kaspar030 commented:
> @kaspar030 I see you pushed a lot of changes, should I re-test? or are more changes coming?

Yes, please re-test! I'll maybe be pushing documentation fixes.

fjmolinas commented on examples/suit_update/README.md:
```diff
@@ -529,3 +510,23 @@ suit/genkey: this recipe generates a ed25519 key to sign the manifest
+
+- compile the application: 
+
+    $ make -Cexamples/suit_update make clean all flash -j8
```
wow wow wow -j8 seems like a bit much, missing space in `-Cexamples/suit_update `

fjmolinas replied:
also `twice in the command`, this should do:

`$ make -C examples/suit_update clean all flash -j`

kaspar030 replied:
> wow wow wow -j8 seems like a bit much, missing space in `-Cexamples/suit_update `

fixed, very used to type "-j8" on my laptop... space is not necessary, i never use it. anyhow, added them where missing for consistency.

kaspar030 replied:
> also `twice in the command`, this should do:

removed the extra make

kaspar030 commented:
I pushed a small for for the Makefile, "$SUIT_KEY" was used instead of "$SUIT_SEC", in "TEST_FILES_EXTRA". Doesn't concern local testing, only "make test-murdock" was broken.

aabadie commented:
What is this PR about exactly ? Does it contain the branch that will actually be PRed upstream ?

If not, I suggest opening small PRs against suit-dev so we can review them. I don't like last minute changes like this. Even if they are *just* documentation fixes.

fjmolinas commented on dist/tools/ethos/setup_network.sh:
```diff
@@ -0,0 +1,49 @@
+TAP=$1
+PREFIX=$2
+_USER=$3
+UHCPD="../uhcpd/bin/uhcpd"
```
This does not work if its not called from `/dist/tools/ethos/`, change for `UHCPD="$(readlink -f $(dirname $0)"/../uhcpd/bin")/uhcpd"`?

kaspar030 replied:
good point. changed to ```: ${UHCPD:="$(readlink -f $(dirname $0)"/../uhcpd/bin")/uhcpd"}``` so it can still be overridden.

fjmolinas reviewed (changes requested):
Some minor comments, otherwise works like a charm and I like the approach.

aabadie commented on examples/suit_update/README.md:
```diff
@@ -167,6 +168,15 @@ In this case the RIOT node can be reached from the host using its global address
 `aiocoap-fileserver` is used for hosting firmwares available for updates. Device
 retrieve the new firmware using the CoAP protocol.
 
+*either*:
```
This seems unrelated to the goal of this PR, right ?

fjmolinas replied:
No, the lower ~~commit~~edits say fileserver needs to be added it to $PATH and this is then used in the test script to setup the fileserver.

aabadie replied:
How is this useful from the user point of view ? except by introducing confusion ? I mean this way of installing is only done for the CI so there's no need to document it here. Or maybe only give one way of installing the fileserver ?

fjmolinas replied:
Hmm you are right, maybe it should be split in test documentation?

kaspar030 replied:
I changed the prerequisites section to include the wget method, saving to ```~/.local/bin```.
Also added a paragraph with the requirement to add "~/.local/bin" to PATH.

aabadie commented on LOSTANDFOUND.md:
```diff
@@ -100,3 +100,14 @@ Author(s):
 - code broken and excluded from all tests for a long time
 - no maintainer available
+
+### gnrc_pktbuf_duplicate_upto() [b83430aa625a1d42f11f9badf5e5cfbb8efacd99]
```
Looks like there's some rebase issue with this PR.

kaspar030 replied:
it is just that the PR shows the new commits in master that are not yet in suit-dev

kaspar030 committed:
examples/suit_update: fix TEST_EXTRA_FILES SUIT_KEY -> SUIT_SEC

kaspar030 committed:
fixup! dist/tools/ethos: add setup_network.sh script

kaspar030 committed:
examples/suit_update: update README.md

kaspar030 committed:
DEPENDENCY MERGE: pr/always_flash_riotboot_if_required

kaspar030 committed:
DEPENDENCY MERGE: pr/murdock_allow_multiple_files_for_test_job

kaspar030 committed:
DEPENDENCY MERGE: add_riotboot_test_script

kaspar030 committed:
DEPENDENCY MERGE: suit-dev

kaspar030 commented:
> What is this PR about exactly ? Does it contain the branch that will actually be PRed upstream ?
 
This is my WIP branch that integrates suit-dev and the required upstream PRs, and adds the automatic test on top.

> If not, I suggest opening small PRs against suit-dev so we can review them. 

Not possible, as suit-dev is missing dependencies.

kaspar030 committed:
fixup! examples/suit_update: update README.md

kaspar030 committed:
fixup! fixup! examples/suit_update: update README.md

kaspar030 committed:
fixup! examples/suit_update: drop APP_VER?=0 requirement of test script

kaspar030 commented:
- rebased all branches
- fixed uhcp output messing up the slot printout by guarding with irq_disable()
- addressed @fjmolinas comments

I propose I create the squashed master-PR branch, we all give it a last look, and if everyone's happy, I open the PR?

fjmolinas commented:
> I propose I create the squashed master-PR branch, we all give it a last look, and if everyone's happy, I open the PR?

Lets do that, are you ok with following the procedure that I sent on the mailing list?

kaspar030 commented:
> Lets do that, are you ok with following the procedure that I sent on the mailing list?

didn't get that, when did you send it?

kaspar030 commented:
> didn't get that, when did you send it?

ah got it, sorry

fjmolinas commented:
Should we close this one?
